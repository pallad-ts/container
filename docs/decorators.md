# Decorators
Creating service definition via container (`definitionWithConstructor` etc) might be a tedious and error-prone work.
That's why _alpha-dic_ supports decorators for classes.

NOTE! Hopefully this is obvious but Decorators are not supported by node.js natively. You need to use Typescript (highly recommended) or other transpiler (like Babel) to make it work.

```typescript
const {Service, Annotation, Inject} = require('alpha-dic');

@Service() // you can also provide a name @Service('name'). By default class/function name inferred.
@Annotation({name: 'extraAnnotation'})
export class NotificationCenter {
    
    @Inject('superExtraService')
    public extraService: ExtraService;
    
    constructor(
        @Inject('EmailRenderer') emailRenderer: EmailRenderer, 
        @Inject('EmailSender') emailSender: EmailSender) {
    }
}

```

## Setup

Since decorator is just regular function you need to give it an opportunity to get called. To do soo you can use `loadServices` function that loads all modules matching given patterns.  
```typescript
import {createStandard, loadServices} from 'alpha-dic';

const container = createStandard();

// loads all files (with extensions .ts, .js, .tsx but ignores .d.ts, .d.tsx) in given directory so all @Service decorator calls got a chance to register service definitions
loadServices(container, {
    patterns: './services/*.*' // note ".*" to make sure glob can catch all files
}); 
```