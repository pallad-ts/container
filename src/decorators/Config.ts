import {ConfigRequestArg} from "../args/ConfigRequestArg";
import {ensureMetadata} from "../serviceMetadata";

export function Config(path: string, defaultValue?: any): ParameterDecorator & PropertyDecorator;
export function Config(...args: [string, undefined | any]): ParameterDecorator & PropertyDecorator {
    const request = ConfigRequestArg.create(...args);

    return function (target: any, property: string | symbol, indexOrDescriptor?: number | TypedPropertyDescriptor<any>) {
        const isParameterDecorator = typeof indexOrDescriptor === 'number';
        if (isParameterDecorator) {
            ensureMetadata(target).constructorArguments[indexOrDescriptor as number] = request;
        } else {
            ensureMetadata(target.constructor).propertiesInjectors.set(property, request);
        }
    }
}