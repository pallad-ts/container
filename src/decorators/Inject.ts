import {ContainerArg} from "../ContainerArg";
import 'reflect-metadata';
import {Reference} from "../Reference";
import * as is from 'predicates';
import {ensureMetadata} from "./serviceMetadata";


const assertServiceNameOrContainerArg = is.assert(
    is.any(is.string, is.instanceOf(ContainerArg)),
    '@Inject argument must be a string that represents service name or an object of ContainerArg instance'
);

export function Inject(serviceName: string | ContainerArg) {
    assertServiceNameOrContainerArg(serviceName);

    const arg = is.string(serviceName) ? Reference.one.name(serviceName) : serviceName;
    return function (target: any, property: string | symbol, indexOrDescriptor?: number | TypedPropertyDescriptor<any>) {
        const isParameterDecorator = typeof indexOrDescriptor === 'number';
        if (isParameterDecorator) {
            // argument decorator
            ensureMetadata(target).constructorArguments[<number>indexOrDescriptor] = arg;
        } else {
            // property decorator
            ensureMetadata(target.constructor).propertiesInjectors.set(property, arg);
        }
    }
}