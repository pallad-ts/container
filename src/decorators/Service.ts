import {Container} from "../Container";
import {ensureMetadata, createDefinitionFromMetadata} from "../serviceMetadata";

import 'reflect-metadata';
import {Definition} from "../Definition";
import {ServiceName} from "../types";
import {randomName} from "../randomName";

export interface ServiceType {
    (name?: ServiceName): ClassDecorator;

    useContainer(container?: Container): void;

    _container?: Container
}

const DEFINITION_KEY = Symbol('__alphaDic-ServiceDefinition');
export const Service = function (name?: ServiceName) {
    return function (constructor: {new(...args: any[]): any}) {
        const finalName = name || randomName(constructor.name);

        const metadata = ensureMetadata(constructor);
        metadata.name = finalName;

        const definition = createDefinitionFromMetadata(metadata, constructor);
        Reflect.defineMetadata(DEFINITION_KEY, definition, constructor);

        if (Service._container) {
            Service._container.registerDefinition(definition);
        } else if (!process.env.ALPHA_DIC_NO_SERVICE_CONTAINER) {
            console.warn('There is no container registered in @Service decorator. ' +
                'Use Service.useContainer(container) to define the container. ' +
                'Without it service definitions created via @Service decorator cannot be registered automatically. ' +
                'If you wish to register service definitions manually via getDefinitionForClass set ALPHA_DIC_NO_SERVICE_CONTAINER environment variable to disable this warning.')
        }
    }
} as ServiceType;

Service.useContainer = function (container: Container) {
    Service._container = container;
};

export function getDefinitionForClass(clazz: Function): Definition {
    return Reflect.getMetadata(DEFINITION_KEY, clazz);
}
