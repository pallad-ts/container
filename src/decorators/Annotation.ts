import {ensureMetadata} from "./serviceMetadata";
import {getDefinitionForClass} from "./Service";

export function Annotation(annotation: any): ClassDecorator {
    return (clazz: Function) => {
        const definition = getDefinitionForClass(clazz);
        if (definition) {
            definition.annotate(annotation);
        }
        ensureMetadata(clazz).annotations.push(annotation);
    }
}