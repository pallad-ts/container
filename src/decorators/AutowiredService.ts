import {Service} from "./Service";
import {ServiceName} from "../types";
import {ensureMetadata} from "../serviceMetadata";
import * as is from "predicates";
import {TypeRef} from "../TypeRef";
import {errors, ReferenceArg} from "../index";
import {format} from "util";

export function AutowiredService(name?: ServiceName): ClassDecorator {
    return (clazz: Function) => {
        const paramTypes: Function[] = Reflect.getMetadata("design:paramtypes", clazz);
        const metadata = ensureMetadata(clazz);

        if (is.array(paramTypes) && clazz.length > 0) {
            for (const [index, paramType] of paramTypes.entries()) {
                if (metadata.constructorArguments[index]) {
                    continue;
                }
                const ref = TypeRef.createFromType(paramType);
                if (ref === undefined) {
                    throw errors.AUTOWIRING_FAILED(
                        format(errors.AUTOWIRING_FAILED.defaultMessage, `constructor (of ${clazz.name}) argument nr: ${index}`)
                    );
                }
                metadata.constructorArguments[index] = ReferenceArg.one.type(ref);
            }
        } else if (clazz.length > 0) {
            throw errors.AUTOWIRING_NO_METADATA();
        }
        Service(name)(clazz);
    }
}