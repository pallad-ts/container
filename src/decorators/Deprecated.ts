import {deprecated} from "../middlewares/deprecated";
import {Annotation} from "./Annotation";

export function Deprecated(deprecationNote: string): any {
    return function (target: any) {
        Annotation(deprecated(deprecationNote))(target);
    }
}