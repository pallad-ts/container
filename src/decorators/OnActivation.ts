import { Hook, onActivation } from "../";
import { Annotation } from "./Annotation";

// eslint-disable-next-line @typescript-eslint/naming-convention
export function OnActivation(hook: Hook) {
	return (target: any) => {
		Annotation(onActivation(hook))(target);
	};
}
