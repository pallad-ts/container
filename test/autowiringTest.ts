import { Container, Service } from "@src/.";
import { AutowiredService } from "@src/decorators/AutowiredService";
import { Inject } from "@src/decorators/Inject";

describe("autowiring", () => {
	let container: Container;

	const NAME = "example";

	class Foo implements FooInterface {
		// tslint:disable-next-line:no-empty
		bonk() {}
	}

	interface FooInterface {
		bonk(): void;
	}

	class Bar {}

	beforeEach(() => {
		container = new Container();

		container.definitionWithConstructor("foo", Foo);
		container.definitionWithConstructor("bar", Bar);
		Service.useContainer(container);
	});

	afterEach(() => {
		Service._container = undefined;
	});

	it("simple", async () => {
		@AutowiredService(NAME)
		class Example {
			constructor(readonly foo: Foo) {}
		}

		const service = await container.resolve<Example>(NAME);

		expect(service).toBeInstanceOf(Example);

		expect(service.foo).toBeInstanceOf(Foo);
	});

	it("with properties", async () => {
		@AutowiredService(NAME)
		class Example {
			@Inject()
			bar!: Bar;

			@Inject()
			foo!: Foo;
		}

		const service = await container.resolve<Example>(NAME);

		expect(service.bar).toBeInstanceOf(Bar);
		expect(service.foo).toBeInstanceOf(Foo);
	});

	it("nothing to autowire", async () => {
		@AutowiredService(NAME)
		class Example {}

		const service = await container.resolve<Example>(NAME);
		expect(service).toBeInstanceOf(Example);
	});

	describe("overwriting an argument", () => {
		it("with type reference", async () => {
			@AutowiredService(NAME)
			class Example {
				constructor(@Inject(Bar) readonly foo: Foo) {}
			}

			const service = await container.resolve<Example>(NAME);

			expect(service.foo).toBeInstanceOf(Bar);
		});

		it("with explicit reference", async () => {
			@AutowiredService(NAME)
			class Example {
				constructor(@Inject("bar") readonly foo: Foo) {}
			}

			const service = await container.resolve<Example>(NAME);
			expect(service.foo).toBeInstanceOf(Bar);
		});

		it("interface reference", async () => {
			@AutowiredService(NAME)
			class Example {
				constructor(@Inject(Foo) readonly foo: FooInterface) {}
			}

			const service = await container.resolve<Example>(NAME);
			expect(service.foo).toBeInstanceOf(Foo);
		});
	});

	describe("fails", () => {
		it("could not autowire interfaces", () => {
			expect(() => {
				@AutowiredService(NAME)
				class Example {
					constructor(readonly foo: FooInterface) {}
				}
			}).toThrowErrorMatchingSnapshot();
		});

		it("could not use @Inject without arg for arguments", () => {
			expect(() => {
				const Inj = Inject as any;

				@AutowiredService(NAME)
				class Example {
					constructor(@Inj() readonly arg: any) {}
				}
			}).toThrowErrorMatchingSnapshot();
		});

		it("could not autowire union types", () => {
			expect(() => {
				@AutowiredService()
				class Example {
					constructor(readonly arg: Foo | Bar) {}
				}
			}).toThrowErrorMatchingSnapshot();
		});

		it("multiple instances of same type", () => {
			container.definitionWithConstructor("another", Foo);

			@AutowiredService(NAME)
			class Example {
				constructor(readonly foo: Foo) {}
			}

			return expect(container.resolve(NAME)).rejects.toThrowErrorMatchingSnapshot();
		});

		it("explicit referencing reserved type", () => {
			expect(() => {
				@AutowiredService(NAME)
				class Example {
					constructor(@Inject(Object) readonly foo: Foo) {}
				}
			}).toThrowErrorMatchingSnapshot();
		});
	});
});
