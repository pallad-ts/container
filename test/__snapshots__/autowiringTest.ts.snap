// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autowiring fails could not autowire interfaces 1`] = `"Inferred type for constructor (of Example) argument nr: 0 is an interface, union type or built-in object that is not supported"`;

exports[`autowiring fails could not autowire union types 1`] = `"Inferred type for constructor (of Example) argument nr: 0 is an interface, union type or built-in object that is not supported"`;

exports[`autowiring fails could not use @Inject without arg for arguments 1`] = `"Using @Inject decorator without ref for function arguments is prohibited"`;

exports[`autowiring fails explicit referencing reserved type 1`] = `"Inferred type for constructor (of Example) argument nr: 0 is an interface, union type or built-in object that is not supported"`;

exports[`autowiring fails multiple instances of same type 1`] = `"Multiple services found (foo, another) with following lookup: by type: instance of class \\"Foo\\""`;
